File: demo/app/src/main/java/com/timper/lonelysword/app/data/remote/ErrorTransformer.java
package com.timper.lonelysword.app.data.remote;

import io.reactivex.Flowable;
import io.reactivex.FlowableTransformer;
import org.reactivestreams.Publisher;

/**
 * User: tangpeng.yang
 * Date: 2019-08-20
 * Description:
 * FIXME
 */
public class ErrorTransformer<T> implements FlowableTransformer<BaseResponse<T>, T> {


    @Override
    public Publisher<T> apply(Flowable<BaseResponse<T>> upstream) {
        return null;
    }
}

File: demo/module/src/main/java/com/timper/module/data/remote/ErrorTransformer.kt
package com.timper.module.data.remote

import io.reactivex.BackpressureStrategy
import io.reactivex.Flowable
import io.reactivex.FlowableOnSubscribe
import io.reactivex.FlowableTransformer
import java.lang.RuntimeException

/**
 * @author op
 * @version 1.0
 * @description 对response数据进行拦截处理
 * @createDate 2016/3/24
 */
class ErrorTransformer<T> : FlowableTransformer<BaseResponse<T>, T> {

    override fun apply(upstream: Flowable<BaseResponse<T>>): Flowable<T> {
        return upstream.flatMap { it -> flatResponse(it) }
    }

    private fun flatResponse(response: BaseResponse<T>): Flowable<T> {
        return Flowable.create(FlowableOnSubscribe { subscriber ->
            if (!subscriber.isCancelled) {
                when (response.errorCode) {
                    0 -> {
                        response.data?.let {
                            subscriber.onNext(it)
                        }
                        subscriber.onComplete()
                    }
                    else -> {
                        subscriber.onError(RuntimeException(response.errorMsg ?: "加载失败"))
                        return@FlowableOnSubscribe
                    }
                }
            }
        }, BackpressureStrategy.BUFFER)
    }
}
